File: app/src/net/reichholf/dreamdroid/fragment/dialogs/SleepTimerDialog.java
/* © 2010 Stephan Reichholf <stephan at reichholf dot net>
 *
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */

package net.reichholf.dreamdroid.fragment.dialogs;

import android.app.Dialog;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.CheckBox;
import android.widget.RadioGroup;

import androidx.annotation.NonNull;

import com.github.stephenvinouze.materialnumberpickercore.MaterialNumberPicker;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import net.reichholf.dreamdroid.R;
import net.reichholf.dreamdroid.helpers.ExtendedHashMap;
import net.reichholf.dreamdroid.helpers.Python;
import net.reichholf.dreamdroid.helpers.enigma2.SleepTimer;

/**
 * @author sre
 */
public class SleepTimerDialog extends AbstractDialog {
    private static final String KEY_TIMER = "timer";
    private ExtendedHashMap mSleepTimer;

    public static SleepTimerDialog newInstance(ExtendedHashMap sleepTimer) {
        SleepTimerDialog f = new SleepTimerDialog();
        Bundle args = new Bundle();
        args.putSerializable(KEY_TIMER, sleepTimer);
        f.setArguments(args);
        return f;
    }

    @SuppressWarnings("unchecked")
    private void init() {
        mSleepTimer = ((ExtendedHashMap) getArguments().getSerializable(KEY_TIMER)).clone();
    }

    public interface SleepTimerDialogActionListener {
        void onSetSleepTimer(String time, String action, boolean enabled);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        init();
        final View view = LayoutInflater.from(getContext()).inflate(R.layout.sleeptimer, null);
        final MaterialNumberPicker time = view.findViewById(R.id.NumberPicker);
        final CheckBox enabled = view.findViewById(R.id.CheckBoxEnabled);
        final RadioGroup action = view.findViewById(R.id.RadioGroupAction);

        time.setMinValue(0);
        time.setMaxValue(999);

        int min = 90;
        try {
            min = Integer.parseInt(mSleepTimer.getString(SleepTimer.KEY_MINUTES));
        } catch (NumberFormatException nfe) {
            Log.w("bla", nfe.getLocalizedMessage());
        }

        boolean enable = Python.TRUE.equals(mSleepTimer.getString(SleepTimer.KEY_ENABLED));
        String act = mSleepTimer.getString(SleepTimer.KEY_ACTION);

        time.setValue(min);
        enabled.setChecked(enable);

        if (SleepTimer.ACTION_SHUTDOWN.equals(act))
            action.check(R.id.RadioButtonShutdown);
        else
            action.check(R.id.RadioButtonStandby);

        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(getContext());
        builder.setTitle(R.string.sleeptimer)
                .setView(view)
                .setPositiveButton(R.string.save, (dialog, which) -> {
                    MaterialNumberPicker time1 = view.findViewById(R.id.NumberPicker);
                    CheckBox enabled1 = view.findViewById(R.id.CheckBoxEnabled);
                    RadioGroup action1 = view.findViewById(R.id.RadioGroupAction);

                    String t = Integer.valueOf(time1.getValue()).toString();
                    int id = action1.getCheckedRadioButtonId();
                    String a = SleepTimer.ACTION_STANDBY;

                    if (id == R.id.RadioButtonShutdown) {
                        a = SleepTimer.ACTION_SHUTDOWN;
                    }

                    ((SleepTimerDialogActionListener) getActivity()).onSetSleepTimer(t, a, enabled1.isChecked());
                })
                .setNegativeButton(R.string.cancel, (dialog, which) -> dismiss());
        return builder.create();
    }
}

File: app/src/net/reichholf/dreamdroid/fragment/dialogs/SleepTimerDialog.kt
package net.reichholf.dreamdroid.fragment.dialogs

import android.app.Dialog
import android.os.Bundle
import android.util.Log
import android.view.ViewGroup
import androidx.compose.ui.platform.ComposeView
import androidx.lifecycle.setViewTreeLifecycleOwner
import androidx.lifecycle.setViewTreeViewModelStoreOwner
import androidx.savedstate.setViewTreeSavedStateRegistryOwner
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import net.reichholf.dreamdroid.R
import net.reichholf.dreamdroid.helpers.ExtendedHashMap
import net.reichholf.dreamdroid.helpers.Python
import net.reichholf.dreamdroid.helpers.enigma2.SleepTimer
import net.reichholf.dreamdroid.ui.dialogs.SleepTimerUiState
import net.reichholf.dreamdroid.ui.dialogs.bindSleepTimerScreen

/**
 * Sleep timer form. Compose Material 3 body; Save still goes to
 * [SleepTimerDialogActionListener] on the activity.
 */
class SleepTimerDialog : AbstractDialog() {
    private lateinit var uiState: SleepTimerUiState

    interface SleepTimerDialogActionListener {
        fun onSetSleepTimer(time: String, action: String, enabled: Boolean)
    }

    private fun initState() {
        @Suppress("UNCHECKED_CAST")
        val timer = (requireArguments().getSerializable(KEY_TIMER) as ExtendedHashMap).clone() as ExtendedHashMap
        var minutes = 90
        try {
            minutes = Integer.parseInt(timer.getString(SleepTimer.KEY_MINUTES))
        } catch (nfe: NumberFormatException) {
            Log.w(TAG, nfe.localizedMessage ?: "")
        }
        val enabled = Python.TRUE == timer.getString(SleepTimer.KEY_ENABLED)
        val action = timer.getString(SleepTimer.KEY_ACTION) ?: SleepTimer.ACTION_STANDBY
        uiState = SleepTimerUiState(minutes, enabled, action)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        initState()
        val host = this
        val composeView = ComposeView(requireContext()).apply {
            layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT,
            )
            setViewTreeLifecycleOwner(host)
            setViewTreeViewModelStoreOwner(host)
            setViewTreeSavedStateRegistryOwner(host)
            bindSleepTimerScreen(uiState)
        }
        return MaterialAlertDialogBuilder(requireContext())
            .setTitle(R.string.sleeptimer)
            .setView(composeView)
            .setPositiveButton(R.string.save) { _, _ ->
                (activity as SleepTimerDialogActionListener).onSetSleepTimer(
                    uiState.minutes.toString(),
                    uiState.action,
                    uiState.enabled,
                )
            }
            .setNegativeButton(R.string.cancel) { _, _ -> dismiss() }
            .create()
    }

    companion object {
        private const val KEY_TIMER = "timer"
        private const val TAG = "SleepTimerDialog"

        @JvmStatic
        fun newInstance(sleepTimer: ExtendedHashMap): SleepTimerDialog {
            val f = SleepTimerDialog()
            val args = Bundle()
            args.putSerializable(KEY_TIMER, sleepTimer)
            f.arguments = args
            return f
        }
    }
}
